core/register_core_types.cpp

```diff
@@ -266,6 +267,7 @@ void register_core_types() {
 	GDREGISTER_ABSTRACT_CLASS(ResourceImporter);
 
 	GDREGISTER_CLASS(GDExtension);
+	GDREGISTER_CLASS(GodotInstance);
```

Comment:
Actually, thinking about this a little more: Do we even need to have a bound `GodotInstance` class?

There should only every be 1 instance of it. We only want to ever use it from GDExtension (not scripts). It has only 4 methods that we need to call.

Couldn't we instead just add 4 functions to `gdextension_interface.h`?

Maybe we'll still want a class internally, but I don't think we necessarily need to bind a class for this. Not binding it avoids the whole issue with controlling its construction and hiding it from script.

Reply:
This is a bug, we wanted to declare it as GDREGISTER_ABSTRACT_CLASS() like we did with the DisplayServerEmbedded class. It effectively disables the creation function in the API, so it will not be possible to create an object except with the create_instance function.

There are a couple reasons why I would like to keep it as a GDExtension class:
- I see this as a first (tiny) step to turn Godot into a library where you can create multiple instances. Right now, it will be limited to one instance (as you already remarked below), but I see no reason why we shouldn't have an API that could allow it in the future without any changes on the client side.
- Registering it as GDExtension will reduce the burden on the language binding developers: no need to bind it manually, the generator will take care of it. We already demonstrate this for Swift and C++ in the samples.
- My long(er) term idea is that we should change the way Godot is initialized by extending GodotInstance:
  - On create_instance() we would only run a stripped down version of Main::setup() without the command line parsing and configuration.
  - GodotInstance then could provide a Configuration API, to configure the instance before start() is called. Of course we can provide a configure(String[] args) method that will call the command line parser.
  - Then when start() is called, the rest of the startup sequence would be executed.
  - GodotInstance could also add the APIs to register/unregister embedded GDExtensions. In this case we would not have to add separate GDExtension function pointers for these, which would have to be handled by the language binding developers manually.
  - My hope is that this extended API of the GodotInstance class would be used for configuring, starting and controlling the Godot instance on all platforms - either through the GDExtension API or linking directly to it inside the engine. This could also reduce the fragmentation (and the required code) across the platform implementations.

Reply:
Ok, thanks for the explanation!

I'm not totally opposed to registering a `GodotInstance` class. But when we have something we don't want available to scripting and/or want to tightly control, we frequently avoid registering/exposing it, and add functions to the GDExtension interface instead, and this kind of feels like one of those cases.

If we do register the `GodotInstance` class, we just need to make sure it can't be created or used in ways we don't want, and, _especially_ that scripting can't access it.

Reply:
Now GodotInstance is registered as abstract and the only place it is instantiated in libgodot_create_godot_instance()